cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,364 @@
+
+	// Read the current Istio version from the the cluster
+	currentVer, err := retrieveControlPlaneVersion(kubeClient, istioNamespace, l)
+	if err != nil {
+		l.logAndPrintf("Failed to read the current Istio version, error: %v", err)
+		if !args.force {
+			return err
+		}
+	}
+
+	// Read the current Istio installation values from the cluster
+	currentValues, err := readValuesFromInjectorConfigMap(kubeClient, istioNamespace)
+	if err != nil {
+		l.logAndPrintf("Failed to read the current Istio installation values, "+
+			"error: %v", err)
+		if !args.force {
+			return err
+		}
+	}
+
+	// Check if the upgrade currentVer -> targetVersion is supported
+	err = checkSupportedVersions(currentVer, targetVersion, args.versionsURI, l)
+	if err != nil {
+		l.logAndPrintf("Upgrade version check failed: %v -> %v. Error: %v",
+			currentVer, targetVersion, err)
+		if !args.force {
+			return err
+		}
+	} else {
+		l.logAndPrintf("Upgrade version check passed: %v -> %v.\n", currentVer, targetVersion)
+	}
+
+	checkUpgradeValues(currentValues, targetValues, l)
+	waitForConfirmation(args.yes, l)
+
+	// Run pre-upgrade hooks
+	// TODO(elfinhe): add err handling when hook refactoring is done
+	runPreUpgradeHooks(kubeClient, istioNamespace,
+		currentVer, targetVersion, currentValues, targetValues, rootArgs.dryRun, l)
+
+	// Apply the Istio Control Plane specs reading from inFilename to the cluster
+	err = genApplyManifests(nil, args.inFilename, rootArgs.dryRun,
+		rootArgs.verbose, args.kubeConfigPath, args.context, upgradeWaitSecWhenApply, l)
+	if err != nil {
+		l.logAndPrintf("Failed to apply the Istio Control Plane specs. Error: %v", err)
+		return err
+	}
+
+	// Run post-upgrade hooks
+	// TODO(elfinhe): add err handling when hook refactoring is done
+	runPostUpgradeHooks(kubeClient, istioNamespace,
+		currentVer, targetVersion, currentValues, targetValues, rootArgs.dryRun, l)
+
+	if !args.wait {
+		l.logAndPrintf("Upgrade submitted. Please use `istioctl version` to check the current versions.")
+		return nil
+	}
+
+	// Waits for the upgrade to complete by periodically comparing the each
+	// component version to the target version.
+	err = waitUpgradeComplete(kubeClient, istioNamespace, targetVersion, l)
+	if err != nil {
+		l.logAndPrintf("Failed to wait for the upgrade to complete. Error: %v", err)
+		return err
+	}
+
+	// Read the upgraded Istio version from the the cluster
+	upgradeVer, err := retrieveControlPlaneVersion(kubeClient, istioNamespace, l)
+	if err != nil {
+		l.logAndPrintf("Failed to read the upgraded Istio version. Error: %v", err)
+		return err
+	}
+
+	l.logAndPrintf("Success. Now the Istio control plane is running at version %v.", upgradeVer)
+	return nil
+}
+
+// checkUpgradeValues checks the upgrade eligibility by comparing the current values with the target values
+func checkUpgradeValues(curValues string, tarValues string, l *logger) {
+	diff := compare.YAMLCmp(curValues, tarValues)
```

Comment:
Could you please raise an issue that we need to exclude the overridden values? I think the text below is ok for now but it's hard for the average user to understand what's going on.

Reply:
Issue created: https://github.com/istio/istio/issues/17916